packages/react/src/ui/DatePickerPopup/OneDatePickerPopup.tsx

```diff
@@ -0,0 +1,159 @@
+import { Button } from "@/components/Actions/Button"
+import {
+  GranularityDefinitionKey,
+  OneCalendar,
+  granularityDefinitions,
+} from "@/experimental/OneCalendar"
+import { DateRange } from "@/experimental/OneCalendar/types"
+import { ChevronLeft } from "@/icons/app"
+import { Popover, PopoverContent, PopoverTrigger } from "@/ui/popover"
+import { useMemo, useState } from "react"
+import { GranularitySelector } from "./components/GranularitySelector"
+import { PresetList } from "./components/PresetList"
+import { DatePickerValue, DatePreset } from "./types"
+export interface OneDatePickerPopupProps {
+  onSelect?: (value: DatePickerValue | undefined) => void
+  value?: DatePickerValue
+  defaultValue?: DatePickerValue
+  presets?: DatePreset[]
+  granularities?: GranularityDefinitionKey[]
+  minDate?: Date
+  maxDate?: Date
+  disabled?: boolean
+  hideGoToCurrent?: boolean
+  children: React.ReactNode
+  open?: boolean
+  onOpenChange?: (open: boolean) => void
+}
+
+const presetCustom = "__custom__"
+
+export function OneDatePickerPopup({
+  onSelect,
+  defaultValue,
+  presets = [],
+  granularities = ["day"],
+  children,
+  value,
+  ...props
+}: OneDatePickerPopupProps) {
+  const [localValue, setLocalValue] = useState<DatePickerValue | undefined>(
+    value || defaultValue
+  )
+  const granularityDefinition = useMemo(() => {
+    return granularityDefinitions[localValue?.granularity ?? "day"]
+  }, [localValue?.granularity])
+
+  const calendarMode = useMemo(() => {
+    return (
+      granularityDefinitions[localValue?.granularity ?? "day"].calendarMode ||
```

Comment:
I see we a pattern of trying to access the information from local value, and falling back to "day" if not set.

What do you think to have this more prominent. I don't have a great solution in mind, but I'd definitely extracted the default granularity or maybe calculated the current granularity separately

Reply:
makes sense to avoid repeating the `?? 'day'`. Thanks for the suggestion

Reply:
done